akka-persistence/src/main/scala/akka/persistence/Persistence.scala

```diff
@@ -73,65 +73,113 @@ final class PersistenceSettings(config: Config) {
  * Persistence extension.
  */
 class Persistence(val system: ExtendedActorSystem) extends Extension {
+  import Persistence.ActorMemento
+  import scala.collection.concurrent.TrieMap
+
   private val DefaultPluginDispatcherId = "akka.persistence.dispatchers.default-plugin-dispatcher"
+
   private val config = system.settings.config.getConfig("akka.persistence")
 
   val settings = new PersistenceSettings(config)
 
-  private val snapshotStore = createPlugin("snapshot-store") { _ ⇒
+  private def journalDispatchSelector(klaz: Class[_]): String =
+    if (classOf[AsyncWriteJournal].isAssignableFrom(klaz)) Dispatchers.DefaultDispatcherId else DefaultPluginDispatcherId
+
+  private def snapshotDispatchSelector(klaz: Class[_]): String =
     DefaultPluginDispatcherId
+
+  /** Persistent actors and their journal/snapshot plugins - by persistence id. */
+  private val actorPlugins = TrieMap[String, ActorMemento]()
+
+  /** Discovered persistence journal plugins - by journal plugin id. */
+  private val journalPlugins = TrieMap[String, ActorRef]()
+
+  /** Discovered persistence snapshot store plugins - by snapshot plugin id. */
+  private val snapshotPlugins = TrieMap[String, ActorRef]()
+
+  /** Check for default identity. */
+  private def isDefault(text: String) = text == null || text.length == 0
+
+  /** INTERNAL API. Create persistent actor memento. */
+  private[persistence] def register(actor: Eventsourced): Unit = {
+    journalBy(actor.journalPluginId) // Ensure journal is provisioned. 
+    snapshotStoreBy(actor.snapshotPluginId) // Ensure snapshot store is provisioned.
+    actorPlugins.put(actor.persistenceId, ActorMemento(actor.journalPluginId, actor.snapshotPluginId))
   }
 
-  private val journal = createPlugin("journal") { clazz ⇒
-    if (classOf[AsyncWriteJournal].isAssignableFrom(clazz)) Dispatchers.DefaultDispatcherId
-    else DefaultPluginDispatcherId
+  /** INTERNAL API. Delete persistent actor memento. */
+  private[persistence] def unregister(actor: Eventsourced): Unit = {
+    actorPlugins.remove(actor.persistenceId)
   }
 
   /**
-   * Creates a canonical persistent actor id from a persistent actor ref.
+   * Returns a journal actor identified by `journalPluginId`.
+   * When empty, looks in `akka.persistence.journal.plugin` to find configuration entry path.
+   * When configured, uses `journalPluginId` as absolute path to the journal configuration entry.
+   * Configuration entry must contain few required fields, such as `class`. See `src/main/resources/reference.conf`.
    */
-  def persistenceId(persistentActor: ActorRef): String = id(persistentActor)
+  def journalBy(journalPluginId: String): ActorRef = {
+    val configPath = if (isDefault(journalPluginId)) config.getString("journal.plugin") else journalPluginId
+    // TODO remove synchronized after https://issues.scala-lang.org/browse/SI-7943
+    journalPlugins.synchronized(journalPlugins.getOrElseUpdate(configPath, createPlugin(configPath)(journalDispatchSelector)))
```

Comment:
That is a difficult concurrency problem. The tricky part is to only invoke the creator function once.

We have solved that in the registration of Akka extensions: https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/ActorSystem.scala#L745
That is probably not something we would like to duplicate.

It would be possible to register each journal as an extension, but that would require a `ConcurrentHashMap[String, ExtensionId] here for the mapping between the journalPluginId and ExtensionId instance.
I'm not sure if that is the best solution.